app/scripts/controllers/swaps.js

```diff
@@ -0,0 +1,620 @@
+        clearInterval(intervalId)
+        intervalId = null
+
+        const { swapsState } = this.store.getState()
+        if (swapsState.swapsFeatureIsLive) {
+          this.setSwapsLiveness(false)
+        }
+      }
+    })
+
+    fetchAndSetupInterval()
+  }
+
+  /**
+   * This function should only be called via _setupSwapsLivenessFetching.
+   *
+   * Attempts to fetch the swaps feature liveness flag from our API. Tries
+   * to fetch three times at 5-second intervals before giving up, in which
+   * case the value defaults to 'false'.
+   *
+   * Only updates state if the fetched/computed flag value differs from current
+   * state.
+   */
+  async _fetchAndSetSwapsLiveness () {
+    const { swapsState } = this.store.getState()
+    const { swapsFeatureIsLive: oldSwapsFeatureIsLive } = swapsState
+    let swapsFeatureIsLive = false
+    let successfullyFetched = false
+    let numAttempts = 0
+
+    const fetchAndIncrementNumAttempts = async () => {
+      try {
+        swapsFeatureIsLive = Boolean(await this._fetchSwapsFeatureLiveness())
+        successfullyFetched = true
+      } catch (err) {
+        log.error(err)
+        numAttempts += 1
+      }
+    }
+
+    await fetchAndIncrementNumAttempts()
+
+    // The loop conditions are modified by fetchAndIncrementNumAttempts.
+    // eslint-disable-next-line no-unmodified-loop-condition
+    while (!successfullyFetched && numAttempts < 3) {
+      await new Promise((resolve) => {
+        setTimeout(resolve, 5000) // 5 seconds
+      })
+      await fetchAndIncrementNumAttempts()
+    }
+
+    if (!successfullyFetched) {
+      log.error('Failed to fetch swaps feature flag 3 times. Setting to false and trying again next interval.')
+    }
+
+    if (swapsFeatureIsLive !== oldSwapsFeatureIsLive) {
+      this.setSwapsLiveness(swapsFeatureIsLive)
+    }
+  }
+
+  async _modifyValuesForMaxEthMode (newQuotes, accountBalance) {
+    const {
+      swapsState: { customGasPrice },
+    } = this.store.getState()
+
+    const usedGasPrice = customGasPrice || await this._getEthersGasPrice()
+
+    const mappedNewQuotes = mapValues(newQuotes, (quote) => {
+      const oldSourceAmount = quote.sourceAmount
+
+      const gasTotalInWeiHex = calcGasTotal((new BigNumber(quote.maxGas, 10)).toString(16), usedGasPrice)
+      const newSourceAmount = (new BigNumber(accountBalance, 16)).minus(gasTotalInWeiHex, 16).toString(10)
+
+      const newOldRatio = (new BigNumber(newSourceAmount, 10)).div(oldSourceAmount, 10)
+      const oldNewDifference = (new BigNumber(oldSourceAmount, 10)).minus(newSourceAmount, 10)
+
+      const oldDestinationAmount = quote.destinationAmount
+      const newDestinationAmount = (new BigNumber(oldDestinationAmount, 10)).times(newOldRatio)
+
+      const oldValue = quote.trade.value
+      const newValue = (new BigNumber(oldValue, 16)).minus(oldNewDifference, 10)
+
+      return {
+        ...quote,
+        trade: {
+          ...quote.trade,
+          value: newValue,
```

Comment:
Note: this can be a negative number. This may already be handled by the UI logic - I'm not sure yet.